lib/ObjWriter/objwriter.cpp

```diff
@@ -249,6 +252,58 @@ extern "C" void FinishObjWriter(ObjectWriter *OW) {
   delete OW;
 }
 
+extern "C" void CreateDataSection(ObjectWriter *OW, const char *SectionName,
+                                  bool ReadOnly)
+{
+  assert(OW && "ObjWriter is null");
+  Triple TheTriple(TripleName);
+  auto *AsmPrinter = &OW->getAsmPrinter();
+  auto &OST = *AsmPrinter->OutStreamer;
+  MCContext &OutContext = OST.getContext();
+
+  std::string SectionNameStr(SectionName);
+  assert(OW->CustomSections.find(SectionNameStr) != OW->CustomSections.end() && 
+    "Section with duplicate name already exists");
+
+  MCSection *Section = nullptr;
+  SectionKind Kind = ReadOnly ? 
+      SectionKind::getReadOnly() : SectionKind::getData();
+
+  switch (TheTriple.getObjectFormat()) {
+    case Triple::MachO:
+      Section = OutContext.getMachOSection("__DATA", SectionName, 0, Kind);
+      break;
+    case Triple::COFF:
+      {
+        unsigned Characteristics = COFF::IMAGE_SCN_CNT_INITIALIZED_DATA |
+            COFF::IMAGE_SCN_MEM_READ;
+
+        if (!ReadOnly)
+        {
+          Characteristics |= COFF::IMAGE_SCN_MEM_WRITE;
+        }
+        Section = OutContext.getCOFFSection(SectionName, Characteristics, Kind);
+        break;
+      }
+    case Triple::ELF:
+        {
+          unsigned Flags = ELF::SHF_ALLOC;
+          if (!ReadOnly)
+          {
+            Flags |= ELF::SHF_WRITE;
+          }
+          Section = OutContext.getELFSection(SectionName, 
+              ELF::SHT_PROGBITS, Flags);
+          break;
+        }
+    case Triple::UnknownObjectFormat:
```

Comment:
I was following the LLVM convention (e.g: MCObjectFileInfo::InitMCObjectFileInfo) but I can switch this to default if you prefer.